apps/remote_control/lib/lexical/remote_control/search/store/backends/cub/state.ex

```diff
@@ -0,0 +1,242 @@
+
+  The schema presented here does a couple interesting things in order to be efficient and easy-ish
+  to work on. We need to do the following things:
+
+    1. Keep a set of entities organized by their subject, type and subtype.
+    2. Keep a set of entities organized by their reference (which is effectively an id)
+    3. Be able to delete all entities that belong to a given path
+
+  To that end, we store whole entities under the `subject_key` and `ref_key`, which leads to some
+  duplication, but improves query performance, as there's no indirection when we do a subject
+  search.
+
+  We store both `ref_key` and `subject_key` entries under the path_key entry. This allows us to
+  delete entries by path fairly efficiently, as we've enumerated every key that contains a given path,
+  and since `subject_keys` contain a path, only those subjects in a path will be deleted.
+  """
+  alias Lexical.Project
+  alias Lexical.RemoteControl.Search.Indexer.Entry
+  alias Lexical.VM.Versions
+
+  import Record
+
+  defrecord :path_key, sort_key: 1, path: nil
+
+  defrecord :subject_key, type: nil, subtype: nil, subject: nil
+
+  defrecord :ref_key, id: nil
+
+  defstruct project: nil, leader?: false, leader_pid: nil, cub_pid: nil
+
+  # Erlang's total ordering
+  # number < atom < reference < fun < port < pid < tuple < map < nil < list < bit string
+
+  def new_leader(%Project{} = project, cub_pid) do
+    %__MODULE__{
+      project: project,
+      leader_pid: self(),
+      leader?: true,
+      cub_pid: cub_pid
+    }
+  end
+
+  def new_follower(%Project{} = project, leader_pid) do
+    %__MODULE__{
+      leader_pid: leader_pid,
+      leader?: false,
+      project: project
+    }
+  end
+
+  def prepare(%__MODULE__{leader?: true} = state) do
+    case CubDB.size(state.cub_pid) do
+      0 -> {:ok, :empty}
+      _n -> {:ok, :stale}
+    end
+  end
+
+  def prepare(%__MODULE__{}) do
+    {:ok, :stale}
+  end
+
+  def insert(%__MODULE__{} = state, entries) do
+    CubDB.put_multi(state.cub_pid, entries_to_keys_and_values(entries))
+  end
+
+  def drop(%__MODULE__{} = state) do
+    CubDB.clear(state.cub_pid)
+  end
+
+  def replace_all(%__MODULE__{} = state, entries) do
+    keys_and_values = entries_to_keys_and_values(entries)
+
+    CubDB.transaction(state.cub_pid, fn transaction ->
+      CubDB.Tx.clear(transaction)
+
+      transaction =
+        Enum.reduce(keys_and_values, transaction, fn {key, value}, transaction ->
+          CubDB.Tx.put(transaction, key, value)
+        end)
+
+      {:commit, transaction, :ok}
+    end)
+  end
+
+  def delete_by_path(%__MODULE__{} = state, path) do
+    CubDB.transaction(state.cub_pid, fn transaction ->
+      path_key = path_key(path: path)
+
+      {subject_keys, ref_keys} =
+        transaction
+        |> CubDB.Tx.get(path_key, [])
+        |> Enum.unzip()
+
+      deleted_entries =
+        ref_keys
+        |> Enum.map(&CubDB.Tx.get(transaction, &1))
+        |> Enum.to_list()
```

Comment:
I think the `Enum.to_list` is redundant.

```suggestion
      deleted_entries = Enum.map(ref_keys, &CubDB.Tx.get(transaction, &1))
```